File: C/huffman/huffman_node.c
#include <stdlib.h>
#include <stdbool.h>
#include <stdio.h>
#include "huffman_node.h"
#include "../decompression/bytes_vector/bytes_vector.h"

struct huffman_node {
    char letter;
    int weight;
    huffman_node_t *left;
    huffman_node_t *right;
};


huffman_node_t *huffman_node_create(char letter, int frequency) {
    huffman_node_t *node = malloc(sizeof(huffman_node_t));
    if (!node) {
        return NULL;
    }

    node->letter = letter;
    node->weight = frequency;
    node->left = NULL;
    node->right = NULL;

    return node;
}

void huffman_node_destroy(huffman_node_t *node) {
    if (node->left != NULL) {
        huffman_node_destroy(node->left);
    }
    if (node->right != NULL) {
        huffman_node_destroy(node->right);
    }

    free(node);
}

huffman_node_t *huffman_node_merge_nodes(huffman_node_t *node1, huffman_node_t *node2) {
    int w1 = node1->weight;
    int w2 = node2->weight;
    huffman_node_t *left;
    huffman_node_t *right;

    // lower to left, higher to right
    if (w1 < w2) {
        left = node1;
        right = node2;
    } else {
        left = node2;
        right = node1;
    }

    huffman_node_t *new_node = huffman_node_create(0, w1 + w2);
    if (new_node == NULL) {
        return NULL;
    }

    new_node->left = left;
    new_node->right = right;
    return new_node;
}

bool huffman_node_is_leaf(const huffman_node_t *node) {
    return node->left == NULL && node->right == NULL;
}

/*
 * Given a code, it returns it's associated letter. If the code corresponds to a non-leaf node
 * (therefore it does not contain a letter) then it returns -1.
 */
char huffman_node_get_letter(const huffman_node_t* self, const char* tree_code_it) {
    if (huffman_node_is_leaf(self)) {
        return self->letter;
    } else {
        if (*tree_code_it != '\0') {
            if (*tree_code_it == '0') {
                return huffman_node_get_letter(self->left, ++tree_code_it);
            } else { //'1'
                return huffman_node_get_letter(self->right, ++tree_code_it);
            }
        } else {
            return -1;
        }
    }
}

/*
 * Recursively travels the tree in a CLR (center-left-right) manner until it finds (if it exists)
 * the node with the given letter, returning the code associated with the tree path. If the letter
 * is not present in any node, then it returns NULL. If the letter was found it will return an empty string,
 * subsequently adding the corresponding path to it for each recursive call.
 */
bytes_vector_t* huffman_node_get_code(huffman_node_t* self, char letter) {
    if (self->letter == letter) {
        bytes_vector_t* str = (bytes_vector_t*)calloc(1, sizeof(bytes_vector_t));
        if (!str || bytes_vector_init(str)) {
            fprintf(stderr, "Could not allocate memory for the code path!");
            exit(-1);
        }
        return str;
    } else if (huffman_node_is_leaf(self)) {
        return NULL;
    } else {
        bytes_vector_t* left_code = huffman_node_get_code(self->left, letter);
        if (!left_code) {
            bytes_vector_t* right_code = huffman_node_get_code(self->right, letter);
            if (!right_code) return NULL;
            bytes_vector_add_byte(right_code, '1');
            return right_code;
        } else {
            bytes_vector_add_byte(left_code, '0');
            return left_code;
        }
    }
}

void huffman_node_print(huffman_node_t *node) {
    if (huffman_node_is_leaf(node)) {
        printf("leaf: '%c' -> %d\n", node->letter, node->weight);
    } else {
        printf("not leaf: %d\n", node->weight);
    }

    if (node->left != NULL) {
        huffman_node_print(node->left);
    }
    if (node->right != NULL) {
        huffman_node_print(node->right);
    }
}

int huffman_node_weight(const huffman_node_t *node) {
    return node->weight;
}

File: Rust/src/huffman/huffman_node.rs
pub struct HuffmanNode {
    letter: char,
    weight: i32,
    left: Option<Box<HuffmanNode>>,
    right: Option<Box<HuffmanNode>>,
}

pub struct BadPathError;

impl HuffmanNode {

    pub fn new_leaf(letter: char, frequency: i32) -> Box<HuffmanNode> {
        let node = HuffmanNode {
            letter,
            weight: frequency,
            left: None,
            right: None,
        };
        return Box::new(node);
    }

    pub fn merge_nodes(node1: Box<HuffmanNode>, node2: Box<HuffmanNode>) -> Box<HuffmanNode> {
        let w1 = node1.weight;
        let w2 = node2.weight;
        let left;
        let right;

        // lower to left, higher to right
        if w1 < w2 {
            left = node1;
            right = node2;
        } else {
            left = node2;
            right = node1;
        }

        let new_node = HuffmanNode {
            letter: char::from(0),
            weight: w1 + w2,
            left: Some(left),
            right: Some(right),
        };

        return Box::new(new_node);
    }

    pub fn is_leaf(&self) -> bool {
        return match (&self.left, &self.right) {
            (None, None) => { true }
            (_, _) => { false }
        };
    }

    /*
     * Given a code, it returns it's associated letter. If the code corresponds to a non-leaf node
     * (therefore it does not contain a letter) then it returns None.
     */
    pub fn get_letter(&self, tree_code_it: &mut std::str::Chars) -> Option<char> {
        if self.is_leaf() {
            Some(self.letter)
        } else {
            let it = tree_code_it.next();
            if it? == '0' {
                self.left.as_ref().unwrap().get_letter(tree_code_it) //unwrap will never panic because we know for a fact there is a child since we are not a leaf node
            } else { //'1'
                self.right.as_ref().unwrap().get_letter(tree_code_it)
            }
        }
    }

    /*
     * Recursively travels the tree in a CLR (center-left-right) manner until it finds (if it exists)
     * the node with the given letter, returning the code associated with the tree path. If the letter
     * is not present in any node, then it returns an InexistentLetterError.
     */
    pub fn get_code(&self, letter: char) -> Result<String, BadPathError> {
        return if self.letter == letter {
            Ok(String::new())
        } else if self.is_leaf() {
            Err(BadPathError)
        } else {
            let left_code = self.left.as_ref().unwrap().get_code(letter); //it will never panic because we are not a leaf node
            return match left_code {
                Err(_) => {
                    let mut right_code = self.right.as_ref().unwrap().get_code(letter)?;
                    right_code.push('1');
                    Ok(right_code)
                },
                Ok(mut c) => {
                    c.push('0');
                    Ok(c)
                },
            }
        }
    }

    pub fn weight(&self) -> i32 {
        return self.weight;
    }

    pub fn print(&self) {
        if self.is_leaf() {
            println!("leaf: '{}' -> {}", self.letter.escape_default(), self.weight);
        } else {
            println!("not leaf: {}", self.weight);
            match &self.left {
                None => {}
                Some(node) => { node.print() }
            }
            match &self.right {
                None => {}
                Some(node) => { node.print() }
            }
        }
    }
}
